### ISSUE TITLE
[iOS] Support open with

### ISSUE THREAD

**Issue by HarelM** [IsraelHikingMap/Site#1196]
## Feature
Currently only Android supports open-with/share with due to CI limitation of the plugin that I have chose to use:
https://github.com/j3k0/cordova-plugin-openwith/issues/82
So I'm adding this plugin only when building Android.
I have looked for an iOS only alternative and it seems that the other plugin has this issue too:
https://github.com/EternallLight/cordova-plugin-openwith-ios/issues/23
And so I have written some comments and found out there's another fork that should work for iOS and Android and CI:
https://github.com/missive/cordova-plugin-openwith
I'll need to give it a go.

A few users have already encountered issues with the fact that this is not part of the app, which raises the priority of this issue.

**Comment by HarelM** [comment]
@EtienneLem Thanks for the info in the issue there.
I'll be happy to update the readme of your project with stuff that I find that are incorrect/confusing.
Let me know if you want a pull request or that I just write it here, I'll start here so I won't forget.
The first few comments:
1. Installation can't be made using the following line:
`cordova plugin add cordova-plugin-openwith --variable IOS_URL_SCHEME=cordovaopenwithdemo` as there's no plugin with this name in npm.
Probably need a git source until this is released to npm.
2. Also this part is confusing:
"This is a bit modified version of cordova-plugin-openwith by Jean-Christophe Hoelt for iOS"
Since you are using a fork of the cordova-plugin-openwith-ios I thought it doesn't support android although it is written there multiple times (it might just be me though).
3. When installing with the suggested line above (and git) I'm getting: Failed to install 'cordova-plugin-openwith': Error: Variable(s) missing: IOS_BUNDLE_IDENTIFIER
This error is clear, but the documentation should have the example with it
4. I worked hard to get a lot of files working in Android, but I'm not sure how to do it on iOS, worth adding a file type in iOS and Android as an example, see my intent config here:
https://github.com/IsraelHikingMap/Site/blob/master/IsraelHiking.Web/config.xml#L34
I don't have a clue on how to translate it to iOS plist, but I guess I'll need to learn...
5. My build still fails with the following error message:
```
fatal error: 'ShareViewController.h' file not found
#import "ShareViewController.h"
```
Which I think I had with the original plugin as well :-(
See here: https://github.com/j3k0/cordova-plugin-openwith/issues/82
Build output can be seen here:
https://ci.appveyor.com/project/IsraelHikingHost/site/builds/31830135/job/05c7ump151duw7l6

**Comment by HarelM** [comment]
@EtienneLem any updates on this? I'm not sure what I'm doing wrong... :-(

**Comment by EtienneLem** [comment]
This is maybe fixed with https://github.com/missive/cordova-plugin-openwith/commit/17008a89cb269b101e6cd0e2d969a8cf3b10b227.

**Comment by HarelM** [comment]
@EtienneLem Thanks for the update! I'll try to use the latest version you mentioned in a few days and report back here what had changed and if this was successful or not.
Thanks again!

**Comment by HarelM** [comment]
@EtienneLem I have created another build with the latest changes in tour branch adn now I'm getting the following error:
```
error: "ShareExt" requires a provisioning profile with the App Groups feature. Select a provisioning profile in the Signing & Capabilities editor. (in target 'ShareExt' from project 'Israel Hiking Map')
```
Build can be found here:
https://ci.appveyor.com/project/IsraelHikingHost/site/builds/32231667/job/65e54xvaj8vewcsx

The relevant commit can be seen above and is only changing the package.lock file obviously.

**Comment by EtienneLem** [comment]
Right. Cordova really isn’t friendly with multi-targets. We are actually building the iOS app with [Fastlane](https://fastlane.tools) and can set the proper provisioning profiles that way.

So we use `cordova prepare` to initialize all the plugins / xcode, but instead of doing `cordova compile`, we use `fastlane`.

**Comment by HarelM** [comment]
@EtienneLem I was hoping this is handled by the plugin code as well :-(
I was able to sign my app in appveyor CI (took me a while to get it right I must admit) and was hoping to avoid adding another dependency like fastlane.
Having said that, when I was researching the iOS app signing all the answers was to use fastlane, so I probably need to embrace the fact that I need another tool...
If you have an example repo that uses faslane let me know so I'll be able to reduce some of the effort of understanding what I need to configure in order to make fastlane up and running...

**Comment by valleyofdawn** [comment]
Was it resolved successfully or given up (my iphone-using friends want to know)

**Comment by HarelM** [comment]
The status of this issue is open, so this was not resolved. I hope to get to it someday...

**Comment by valleyofdawn** [comment]
I see. What is the recommended way to share routes with iOS IHM app users in the meantime?

**Comment by HarelM** [comment]
Either using files or using the PC to open a share and re-share it with your account, I guess...

**Comment by valleyofdawn** [comment]
That's quite involved. Is there a way to open a URL from within the app, or could you arrange one?

**Comment by HarelM** [comment]
No. This is also not the place for questions. Please delete your non-technical comments.

**Comment by HarelM** [comment]
[Fastlane](https://docs.fastlane.tools/) on windows:
https://blog.vertica.dk/2019/08/15/automated-android-deployment-with-fastlane-on-windows/
Transporter on windows:
https://help.apple.com/itc/transporteruserguide/en.lproj/static.html#aria-apd28806d5ebc904
This is probably the next step in term of CI and local development...
Relevant issue for questions?
https://github.com/fastlane/fastlane/issues/11687

**Comment by HarelM** [comment]
@BenjaminPoncet I've took the latest code of the fovea open with plugin and I'm still getting the following error:
```
fatal error: 'ShareViewController.h' file not found
#import "ShareViewController.h"
```
Any chance you know how to overcome this?
Build log can be found here:
https://ci.appveyor.com/project/IsraelHikingHost/site/builds/34262543/job/6uakm64ae1eqxq4q
Relevant brach that is begin built is the `fastlane-1266`
Any help would be appreciated... I've been trying for a long time to make this work... :-((

**Comment by HarelM** [comment]
A related article:
https://discuss.bitrise.io/t/unable-to-assign-provisioning-profile-for-app-extension/8481

**Comment by HarelM** [comment]
@yuanjilee I saw a post on fastlane issue with what seems to be the same issue I'm facing. How did you manage to solve the problem? Are you using two provisioning profiles one for the app and the second one for the share extention? When I use two profiles I get a basically the problem described here:
https://stackoverflow.com/questions/55238974/bitrise-ionic-shareext-appex-requires-a-provisioning-profile#

**Comment by HarelM** [comment]
Ok, so an initial commit was merged to master and now the plugin is used in the iOS build in my CI and it completes successfully. yay!
I still don't have a clue how to make it associate GPX files on iOS though... :-(
@EtienneLem @j3k0 @NiklasMerz if you guys can direct me how to do it it would be great as I'm pretty clueless when it comes to iOS, file association and the fact that all this needs to run in a CI server without human intervention...

**Comment by HarelM** [comment]
According to a report from our Facebook group it seems that the app is now listed in some share options but does not open the file when selected.
I need to better understand how fix this.

**Comment by HarelM** [comment]
https://stackoverflow.com/questions/2774343/how-do-i-associate-file-types-with-an-iphone-application
The above might be an easier solution...?

**Comment by HarelM** [comment]
This might be relevant for this issue:
https://github.com/apache/cordova-ios/issues/1020

**Comment by skyus** [comment]
Hi Harel this is takuya write
Your question about how to implement open-with in iOS was quite helpful and I solved it on my own, so I'll share it with you.
This is our app.
https://yudo.page.link/app


# It's pretty addictive.
# Xcode is set to use team provisioning so that provisioning automatically is not a problem.
# Create shareExt bundled with apple developer net.site.app.shareextension
# Create a group in apple developer group.net.site.app.shareextension
# Also, set net.site.app and net.site.app.shareextension to group in apple developer
# The target of config.xml is backward compatible <preference name="deployment-target" value="10.0" />.

!Important!
Install the plugin using git !!!
 ionic cordova plugin add https://github.com/j3k0/cordova-plugin-openwith.git --variable ANDROID_MIME_TYPE="application/gpx+xml" -- variable IOS_URL_SCHEME=appscheme --variable IOS_UNIFORM_TYPE_IDENTIFIER=public.data --variable SHAREEXT_DEVELOPMENT_TEAM=" APPLEDEVELOPERTEAMID" --variable PROVISIONING_PROFILE="xxx-xxxxxxxx-xxxxxx--xxxxx" --variable DISPLAY_NAME="Nmae" --variable IOS_BUNDLE_ IDENTIFIER=net.site.app.shareextension --variable IOS_GROUP_IDENTIFIER=group.net.site.app.shareextension 


DISPLAY_NAME="App" --variable IOS_BUNDLE_IDENTIFIER=net.site.app.shareextension --variable IOS_GROUP_IDENTIFIER=group.net.site.app. shareextension 

---------
plugins/cc.fovea.cordova.openwith/src/ios/ShareExtension/ShareExtension-Info.plist 
Rewrite it to make it XML compatible
---------
<?xml version="1.0" encoding="UTF-8"? >
<!DOCTYPE plist PUBLIC "-//Apple///DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>CFBundleDevelopmentRegion</key
        <string>en</string
        <key>CFBundleDisplayName</key> <string>MyApp</string
        <string>MyApp</string>.
        <key>CFBundleExecutable</key>.
        <string>$(EXECUTABLE_NAME)</string>
        <key>CFBundleIdentifier</key>.
        <string>$(PRODUCT_BUNDLE_IDENTIFIER).shareextension</string>.
        <key>CFBundleInfoDictionaryVersion</key>
        <string>6.0</string>.
        <key>CFBundleName</key>.
        <string>$(PRODUCT_NAME)</string>
        <key>CFBundlePackageType</key> <string>XPC!
        <string>XPC!</string>
        <key>CFBundleShortVersionString</key> <string>4.4.4.</string
        <string>4.4.9</string>!
        <key>CFBundleVersion</key> <string>4.4.9</string
        <string>4.4.9</string> <key>NSExtension</key
        <key>NSExtension</key>.
            <dict>
                <key>NSExtensionAttributes</key> <key>NSExtensionAttributes</key
                <dict>
                    <key>NSExtensionActivationRule</key> <dict
                    <string
                    SUBQUERY (
                        extensionItems,
                        $extensionItem,
                            SUBQUERY (
                                $extensionItem.attachments,
                                $attachment,
                                ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.plain-text" OR
                                ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.jpeg" OR
                                ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.jpeg-2000" OR
                                ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.xml"
                            ). @count == $extensionItem.attachments.@count
                    ). @count == 1
                    </string>.
                </dict>
                <key>NSExtensionMainStoryboard</key>.
                <string>MainInterface</string>.
                <key>NSExtensionPointIdentifier</key> <string>com.apple.share-services</string
                <string>com.apple.share-services</string>.
            </dict>
    </dict>
</plist>
---------


Build


Add importUTI in Info.plist in Xcode
https://stackoverflow.com/questions/31163785/trying-to-define-gpx-document-type-in-xcode-6-4

            <key>UTTypeTagSpecification</key>.
            <dict>
                <key>public.filename-extension</key> <array
                <array>
                    <string>gpx</string>.
                </array>
                <key>public.mime-type</key> <string>application/gpx+xml</string
                <string>application/gpx+xml</string>.
            </dict>


---------
platforms/ios/cordova/lib/projectFile.js rewrite in case of compilation error
---------
    var projectName = fs
    .readdirSync(project_dir)
    .find(d => d.includes(".xcworkspace"))
    .replace(".xcworkspace", "");

    var xcBuildConfiguration = xcodeproj.pbxXCBuildConfigurationSection();
    var plist_file_entry = _.find(xcBuildConfiguration, function(entry) {
        return (
        entry.buildSettings &&
        entry.buildSettings.INFOPLIST_FILE &&
        entry.buildSettings.INFOPLIST_FILE.includes(projectName)
        );
    });
---------

It's a pain in the ass next time, so save the plugin and change that to the installation target

cp plugins-dev/cc.fovea.cordova.openwith  plugins-dev/

---------
plugins-dev/cc.fovea.cordova.openwith/plugin.xml
---------
<! --<string>cc.fovea.openwith.$IOS_URL_SCHEME.shareextension</string>-->
<string>net.site.app.shareextension</string>-->
---------

Be sure to build it in Xcode, or provisioning won't eat it.

**Comment by HarelM** [comment]
Thanks for the info!
I ended up not using this plugin after all since the UX around it is not so good - i.e. the current plugin opens a dialog with an option to see an image you share and the plugin enforces only one type of file that can be used etc.
In the end, it seems that cordova already has some kind of mechanism to support opening files without any plugin and this is what I ended up using, along side `webIntent` in Android.
The problem when using this plugin in CI is that cordova-ios still does not support multiple provisioning profiles and this requires a fork I did for cordova-ios which was not merged unfortunately...
See issue #1391 for the complete solution and the relevant service code which utilizes both `webIntent` and cordova-ios build it `handleOpenURL`.
In my case it proved to be a simple and elegant solution that requires a lot less tinkering with xcode and post cordova build things which is very critical for me since I'm only releasing builds which were built on my CI and never release a build which I prepare locally.

Also I've talked with the guys that sent the following PR which solves some of the issues I've listed above:
https://github.com/missive/cordova-plugin-openwith/pull/2
